src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs

```diff
@@ -465,13 +458,33 @@ private void AdvanceReader(BufferSegment consumedSegment, int consumedIndex, Buf
                     _readerAwaitable.SetUncompleted();
                 }
 
-                while (returnStart != null && returnStart != returnEnd)
+                // Reset and try to pool any used blocks
+                if (returnStart != null)
                 {
-                    returnStart.ResetMemory();
-                    ReturnSegmentUnsynchronized(returnStart);
-                    returnStart = returnStart.NextSegment;
+                    int index = _pooledSegmentCount;
+                    BufferSegment[] pool = _bufferSegmentPool;
+                    while (returnStart != returnEnd)
```

Comment:
This seems more complicated than it was. Is there a reason to change `while (returnStart != null && returnStart != returnEnd)` condition?

Reply:
Idea was to skip this section (reading `index`, `pool`, setting `_pooledSegmentCount`) if nothing to do

However the more common skip should probably be `if (returnStart != returnEnd)` rather than `if (returnStart != null)` ?

Reply:
using the `break;` rather than in as the while condition is its already been pre-checked for the first iteration by the `if`